File: ios/Algo/Features/Chat/Models/Conversation.swift
// Conversation.swift
// Algo
//
// Conversation model representing a chat session, along with
// pagination and list response types for the conversations API.

import Foundation

// MARK: - Conversation

struct Conversation: Codable, Identifiable, Sendable, Equatable {

    let id: String
    let title: String?
    let lastMessage: String?
    let messageCount: Int
    let createdAt: Date
    let updatedAt: Date
    var isArchived: Bool

    init(
        id: String,
        title: String? = nil,
        lastMessage: String? = nil,
        messageCount: Int = 0,
        createdAt: Date = .now,
        updatedAt: Date = .now,
        isArchived: Bool = false
    ) {
        self.id = id
        self.title = title
        self.lastMessage = lastMessage
        self.messageCount = messageCount
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.isArchived = isArchived
    }

    enum CodingKeys: String, CodingKey {
        case id
        case title
        case lastMessage = "last_message"
        case messageCount = "message_count"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case isArchived = "is_archived"
    }
}

// MARK: - Conversation Convenience Properties

extension Conversation {

    /// A display-friendly title, falling back to "New Conversation" if nil.
    var displayTitle: String {
        title ?? "New Conversation"
    }

    /// A formatted relative time string for the last update (e.g. "2 hours ago").
    var relativeTimeString: String {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: updatedAt, relativeTo: .now)
    }

    /// A short preview of the last message, truncated for list display.
    var messagePreview: String? {
        guard let lastMessage, !lastMessage.isEmpty else { return nil }
        if lastMessage.count > 100 {
            return String(lastMessage.prefix(100)) + "..."
        }
        return lastMessage
    }
}

// MARK: - ConversationListResponse

/// Response from GET /chat/conversations.
struct ConversationListResponse: Codable, Sendable {
    let conversations: [Conversation]
    let pagination: Pagination
}

// MARK: - Pagination

struct Pagination: Codable, Sendable, Equatable {
    let page: Int
    let limit: Int
    let total: Int?
    let hasMore: Bool?

    enum CodingKeys: String, CodingKey {
        case page
        case limit
        case total
        case hasMore = "has_more"
    }

    /// Whether there are more pages available.
    var canLoadMore: Bool {
        hasMore ?? false
    }
}

File: android/app/src/main/java/com/algonit/algo/features/chat/data/model/Conversation.kt
package com.algonit.algo.features.chat.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Represents a conversation session between the user and the assistant.
 * Matches the response from GET /chat/conversations.
 */
@Serializable
data class Conversation(
    val id: String,
    val title: String? = null,
    @SerialName("last_message") val lastMessage: String? = null,
    @SerialName("message_count") val messageCount: Int,
    @SerialName("created_at") val createdAt: String,
    @SerialName("updated_at") val updatedAt: String
) {
    /**
     * Returns a display-friendly title, falling back to a truncated
     * version of the last message or a default string.
     */
    val displayTitle: String
        get() = title
            ?: lastMessage?.take(50)?.let { if (it.length == 50) "$it..." else it }
            ?: "New Conversation"
}

/**
 * Paginated list response for conversations.
 */
@Serializable
data class ConversationListResponse(
    val conversations: List<Conversation>,
    val pagination: Pagination
)

/**
 * Pagination metadata included in list responses.
 */
@Serializable
data class Pagination(
    val page: Int,
    val limit: Int,
    val total: Int? = null,
    @SerialName("has_more") val hasMore: Boolean? = null
)

/**
 * Room entity for offline conversation caching.
 */
@androidx.room.Entity(tableName = "conversations")
data class ConversationEntity(
    @androidx.room.PrimaryKey val id: String,
    val title: String? = null,
    @androidx.room.ColumnInfo(name = "last_message") val lastMessage: String? = null,
    @androidx.room.ColumnInfo(name = "message_count") val messageCount: Int,
    @androidx.room.ColumnInfo(name = "created_at") val createdAt: String,
    @androidx.room.ColumnInfo(name = "updated_at") val updatedAt: String
)
